processing/src/main/java/io/druid/segment/incremental/OnheapIncrementalIndex.java

```diff
@@ -237,10 +273,18 @@ protected void concurrentRemove(int offset)
   @Override
   public boolean canAppendRow()
   {
-    final boolean canAdd = size() < maxRowCount;
-    if (!canAdd) {
+    final boolean countCheck = size() < maxRowCount;
+    boolean sizeCheck = true;
+    if (maxBytesInMemory != -1) {
+      sizeCheck = sizeInBytes() < maxBytesInMemory;
+    }
+    final boolean canAdd = countCheck && sizeCheck;
+    if (!countCheck) {
       outOfRowsReason = StringUtils.format("Maximum number of rows [%d] reached", maxRowCount);
     }
+    if (!sizeCheck) {
+      outOfRowsReason = StringUtils.format("Maximum size in bytes [%d] reached", maxBytesInMemory);
```

Comment:
This might replace `outOfRowsReason`. It should be checked that both `countCheck` and `sizeCheck` are `false`.

Reply:
Also, please add a null check for `outOfRowsReason` before allocating it and throw an exception if it's not null.

Reply:
How about reporting both reasons in the case where both checks trip at the same time?

Reply:
yeah this was fuzzy and incomplete, fixed it.